avoid single letter names and code cleanup

werdmann opened the pull request:
#### PR description:

This pull request is following up a request that arose in the context of PR #29626 to rename single letter variables which was deferred to later. It is done now in preparation of further development for cmssw_11_3.
May other variables were renamed as well to improve clarity and readability. No functionality changes have been introduced by this.
There were also some other cosmetic changes concerning comments and white-spaces.
Two unused/obsolete methods were removed from the structs in the header files.
A missing initialization of the variable betastop_ has been added. This only affects the results when Tmin was initialized with an illegal value, which of course should never happen.


#### PR validation:

The code was tested in CMSSW_11_3_0_pre2 against relvals and no difference was found to the original code.
runTheMatrix resulted in no failures.

werdmann committed:
avoid single letter names and code cleanup

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32764/20922

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @werdmann  for master.

It involves the following packages:

RecoVertex/PrimaryVertexProducer


@perrotta, @jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @ebrondol, @mtosi, @dgulhan this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6948be/12589/summary.html
**COMMIT**: 01849129ce26a43cd91423edd71d941843ab0a04
**CMSSW**: CMSSW_11_3_X_2021-01-28-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32764/12589/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716596
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2716567
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented:
@werdmann 
I see that in a few places `__restrict__ ` was added.
Did it have any positive input on vectorization of the code? Did  you check the CPU time used by vertexing?

werdmann commented:
> @werdmann
> I see that in a few places `__restrict__ ` was added.
> Did it have any positive input on vectorization of the code? Did you check the CPU time used by vertexing?
@slava77  I only noticed in the process of synchronizing Z and ZT clustering that one of them used restrict and the other did not.
I will address CPU time in another PR soon.

perrotta commented on RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc:
```diff
@@ -24,7 +24,7 @@ DAClusterizerInZ_vect::DAClusterizerInZ_vect(const edm::ParameterSet& conf) {
 
-  // configurable debug outptut debug output
+  // configurable debug outptut
   verbose_ = conf.getUntrackedParameter<bool>("verbose", false);
```
`verbose_` is only used at https://github.com/cms-sw/cmssw/blob/01849129ce26a43cd91423edd71d941843ab0a04/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc#L50-L68

The same functionality could be replaced, in a more thread compliant way, by using the `DEBUG` / `DEBUG_LEVEL` structure which is used everywhere else in this code

perrotta replied:
The very same apply to the other file DAClusterizerInZT_vect.cc

perrotta commented on RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc:
```diff
@@ -256,15 +252,15 @@ DAClusterizerInZT_vect::track_t DAClusterizerInZT_vect::fill(const vector<reco::
+      t_tkwt = 1. / (1. + local_exp(std::pow(atIP.value() / atIP.error(), 2) -
+                                    std::pow(d0CutOff_, 2)));  // reduce weight for high ip tracks
+      if (edm::isNotFinite(t_tkwt) || t_tkwt < std::numeric_limits<double>::epsilon()) {
+        std::cout << "DAClusterizerInZT_vect.fill rejected track t_tkwt " << t_tkwt << std::endl;
```
Remove all (multi-thread unsafe) `cout`'s from this fill() method: alternatively either replace them with some LogMessage's, or protect them with a `DEBUG` flag

perrotta reviewed (commented):
Since you are cleaning this code, please also take what follows into account.

Moreover:
- The `evalF` method defined in https://github.com/cms-sw/cmssw/blob/01849129ce26a43cd91423edd71d941843ab0a04/RecoVertex/PrimaryVertexProducer/interface/DAClusterizerInZ_vect.h#L201 is never used: can it be removed, then? Otherwise, please at least get rid of the thread-unsafe `cout` in it
- Even though I can't imagine that classes with similar names will exist somewhere else in CMSSW, the include guards should better contain the whole file path name, i.e. (e.g.)
  - `RecoVertex_PrimaryVertexProducer_DAClusterizerInZT_vect_h`
  - `RecoVertex_PrimaryVertexProducer_DAClusterizerInZ_vect_h`

perrotta commented:
The two (longish) methods `DAClusterizerInZ_vect::updateTc` and `DAClusterizerInZ_vect::update` are pretty similar, only differing for  updating `swE` in the "Tc" case, and not in the other one.
Some simplification that avoids copy/pasting, and all related risk of making mistakes during possible updates, could be having a single method with an additional (bool?) parameter that rules the two-lines update of `swE`

werdmann commented:
> The two (longish) methods `DAClusterizerInZ_vect::updateTc` and `DAClusterizerInZ_vect::update` are pretty similar, only differing for updating `swE` in the "Tc" case, and not in the other one.
> Some simplification that avoids copy/pasting, and all related risk of making mistakes during possible updates, could be obtained by having a single method with an additional (bool?) parameter that rules the two-lines update of `swE`

That is of course the better solution for the code. The if statement would be inside the loops. Can someone comment on whether this will somehow impact vectorization?
Thanks

makortel commented:
> The if statement would be inside the loops. Can someone comment on whether this will somehow impact vectorization?

In leading order adding a runtime if inside a loop can prevent vectorization. In sub-leading order compiler may be able to implement the if with mask and continue to vectorize, but it would have to be checked from the generated code. By quick look in this case the "whether or not to update `swE`" probably could be given as a template argument, in which case one could use compile-time `if constexpr` that should lead to same behavior as now with explicit two functions.

perrotta commented:
> > The if statement would be inside the loops. Can someone comment on whether this will somehow impact vectorization?
> 
> In leading order adding a runtime if inside a loop can prevent vectorization. In sub-leading order compiler may be able to implement the if with mask and continue to vectorize, but it would have to be checked from the generated code. By quick look in this case the "whether or not to update `swE`" probably could be given as a template argument, in which case one could use compile-time `if constexpr` that should lead to same behavior as now with explicit two functions.

Thank you @makortel!

Wouldn't that work, perhaps another possibility could be to move the `if` outside the loop, for example:
```C++
    // auto-vectorized
    if (Tc) {
      for (unsigned int k = kmin; k < kmax; ++k) {
        vertices.se[k] += vertices.exp[k] * (tmp_trk_tkwt * o_trk_Z_sum);
        auto w = vertices.rho[k] * vertices.exp[k] * (tmp_trk_tkwt * o_trk_Z_sum * o_trk_dz2);
        vertices.sw[k] += w;
        vertices.swz[k] += w * tmp_trk_z;
      }
   else  {
      for (unsigned int k = kmin; k < kmax; ++k) {
        vertices.se[k] += vertices.exp[k] * (tmp_trk_tkwt * o_trk_Z_sum);
        auto w = vertices.rho[k] * vertices.exp[k] * (tmp_trk_tkwt * o_trk_Z_sum * o_trk_dz2);
        vertices.sw[k] += w;
        vertices.swz[k] += w * tmp_trk_z;
        vertices.swE[k] += w * vertices.exp_arg[k] * obeta;
      }
   }
```
Could this also work (even if more naif)?

makortel commented:
> Could this also work (even if more naif)?

Yes, that should also work (with the price of some code duplication).

werdmann commented:
Thank you for the feedback and apologies for the long silence. I verified that Andreas proposal incurs no measurable cpu time penalty and will implement all suggestions asap.

werdmann committed:
implemented pr recommendations

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32764/21072

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #32764 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6948be/12789/summary.html
**COMMIT**: aa450b5cae69441d17e8fe0f05e4c33cd29f32d3
**CMSSW**: CMSSW_11_3_X_2021-02-08-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32764/12789/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

werdmann committed:
removed bad capture in lambda function

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32764/21074

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #32764 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

perrotta commented on RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc:
```diff
@@ -424,51 +303,63 @@ double DAClusterizerInZ_vect::updateTc(double beta, track_t& gtracks, vertex_t&
+    auto o_trk_sum_Z = 1. / tracks.sum_Z[track_num];
+    auto o_trk_dz2 = tracks.dz2[track_num];
+    auto tmp_trk_z = tracks.zpca[track_num];
     auto obeta = -1. / beta;
```
This line can be moved inside the following loop

werdmann replied:
I guess you mean inside the if(useTc) , outside the loop. 
What's that about the evalF? I thought I had removed it and don't see t anymore.

perrotta replied:
> I guess you mean inside the if(useTc) , outside the loop.

Yes (loop -> block,: sorry for the typo), to avoid computing it when not needed

> What's that about the evalF? I thought I had removed it and don't see t anymore.

Right, my bad: it is not there any more, I probably looked at an outdated version of the code

perrotta reviewed (commented):
Moreover: `evalF` method is not needed anywhere, and it can be removed from DAClusterizerInZ_vect.h

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6948be/12793/summary.html
**COMMIT**: 61315073661f97caa586bb70ab4c7a6116bd97d8
**CMSSW**: CMSSW_11_3_X_2021-02-08-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32764/12793/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 178 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2751765
 - DQMHistoTests: Total failures: 1848
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2749895
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

perrotta commented:
please test

There are differences in wfs 136.793 (only DQM) and 136.874 (also reco outputs, and it suspiciously includes also reco vertices). There were some spurious differences that randomly appeared in some recent PRs (e.g. #32859), and I am relaunching the tests just to check if by chance those differences disappear. On the other hand, the fact that there are non negligible differences in the vertices in 136.874 suggests me that perhaps something slipped out in the cleanup...

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6948be/12821/summary.html
**COMMIT**: 61315073661f97caa586bb70ab4c7a6116bd97d8
**CMSSW**: CMSSW_11_3_X_2021-02-09-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32764/12821/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 546 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2751765
 - DQMHistoTests: Total failures: 4611
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2747132
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

perrotta commented:
> * No significant changes to the logs found
> * Reco comparison results: 546 differences found in the comparisons

Even more differences in the last comparison, as wf 23234 added up...

werdmann commented:
I think the second commit introduced a small difference in behaviour. Some of the updates before merge() calls were normal updates in the previous version of the code, but should have been updateTc(). While unifying the two methods into one with a switch, this was done, hence changing the behaviour slightly.
https://github.com/werdmann/cmssw/blob/61315073661f97caa586bb70ab4c7a6116bd97d8/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc#L858
As a test I modified the flag from true to false, removed the previous update() (which was absent in the previous code, too) and modified the update method to only clear swE when the updateTc flag is true. This appears to restore the previous behavior, however, the code of the pull-request is actually closer to the intended behaviour. Effectively the old code used a Tc estimate made at the previous temperature step, which can be slightly different. The merge() method then estimates the critical temperature of a potentially merged cluster based on this and only merges when it is not too high (he corresponding version of merge in the zt clusterizer does not do that because it would be  much more complicated). The effect should be quite small because it only affects a few cases where the estimated Tc is close to the limit. And although the number of reported failures is large I haven't really seen any substantial changes in the histograms.
So I  think it can be undone, however, the code of the pull-request is actually closer to the intended behaviour. 


![OfflinePV__offlinePrimaryVertices_goodvtxNbr](https://user-images.githubusercontent.com/4669785/107658829-d29f3a00-6c86-11eb-9ea0-9a3082dfdca2.png)
![DQM-goodvtxNbr-1](https://user-images.githubusercontent.com/4669785/107658855-d92db180-6c86-11eb-882e-e10a510ee9bd.png)
![DQM-goodvtxNbr-2](https://user-images.githubusercontent.com/4669785/107658869-dc28a200-6c86-11eb-99de-4004bce83944.png)

perrotta commented:
Thank you @werdmann for your check and explanation. I'm not sure I fully understand it, though.

As far as I can see, with the latest commits you
- In  `DAClusterizerInZT_vect.cc` you moved from `update()` to (the previously non existing) `updateTc()` in five places:
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-2cdb54b03c424f592c065bfb23f6d13bdc99c84b6ac27575a27994a010e106aaR1140
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-2cdb54b03c424f592c065bfb23f6d13bdc99c84b6ac27575a27994a010e106aaR1144
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-2cdb54b03c424f592c065bfb23f6d13bdc99c84b6ac27575a27994a010e106aaR1164
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-2cdb54b03c424f592c065bfb23f6d13bdc99c84b6ac27575a27994a010e106aaR1169
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-2cdb54b03c424f592c065bfb23f6d13bdc99c84b6ac27575a27994a010e106aaR1188
- In `DAClusterizerInZ_vect.cc` you moved from `update()` to the previous `updateTc()` in one place:
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-94d5322fc624da9b8622775c726a9ef9245f224365ff10a6bd172a06ad6bd6e2R858
- In `DAClusterizerInZ_vect.cc` you added one `update()` in a position in which it did not exist before:
  - https://github.com/cms-sw/cmssw/pull/32764/files#diff-94d5322fc624da9b8622775c726a9ef9245f224365ff10a6bd172a06ad6bd6e2R843

This is what I see in the code diffs. And, as far as I understand, it would be enough to:
1) set the flag to "false" in the `update()`'s that were originally not `updateTc()` (the six lines that I pointed out above, five in `DAClusterizerInZT_vect.cc` and one in `DAClusterizerInZ_vect.cc`
2) remove the line https://github.com/cms-sw/cmssw/pull/32764/files#diff-94d5322fc624da9b8622775c726a9ef9245f224365ff10a6bd172a06ad6bd6e2R843 in `DAClusterizerInZ_vect.cc`
to recover the past behaviour.

It seems to me that you are describing some different, and more complicate, change of the current code in order to restore the past behaviour. Could you please post here the exact line changes that you envisage, so that I can try to understand what you did?

By the way: in your figures you mention some "#23234", but as far as I understand you meant this PR instead ("#32764"): correct?

werdmann commented:
You are of course right about the PR number, don't now where I got the wrong one from.
Here is the diff between the PR and the "modified" code that recovers (within the limited testing I did) the previous behaviour (the modified lines are marked with *******)
--- a/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc
+++ b/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc
@@ -359,7 +359,7 @@ double DAClusterizerInZ_vect::update(
     gvertices.se[ivertex] = 0.0;
     gvertices.sw[ivertex] = 0.0;
     gvertices.swz[ivertex] = 0.0;
-    gvertices.swE[ivertex] = 0.0;
+    if(updateTc) gvertices.swE[ivertex] = 0.0; //*********************
   }
 
   // loop over tracks
@@ -840,7 +840,7 @@ vector<TransientVertex> DAClusterizerInZ_vect::vertices(const vector<reco::Trans
   }
 
   thermalize(beta, tks, y, delta_lowT_, rho0);
-  update(beta, tks, y, rho0, true);
+  //update(beta, tks, y, rho0, true); //******************
 
 #ifdef DEBUG
   verify(y, tks);
@@ -855,7 +855,8 @@ vector<TransientVertex> DAClusterizerInZ_vect::vertices(const vector<reco::Trans
   // merge again  (some cluster split by outliers collapse here)
   while (merge(y, tks, beta)) {
     set_vtx_range(beta, tks, y);
-    update(beta, tks, y, rho0, true);
+    //update(beta, tks, y, rho0, true);
+    update(beta, tks, y, rho0, false);  //*******************
   }
 
 #ifdef DEBUG

All updates with a true argument should have been updateTc in the previous version. In even older versions, there was only one update method and it was equivalent to updateTc. In the process of trying to avoid some unnecessary computations, updateTc was created. The main work-loop (thermalize) calls the update (without Tc). Hence, whenever Tc is required, i.e. for split and merge, updateTc must be called after the last update. This was apparently not taken into account in that case. Maybe I had forgotten that merge() uses Tc (it only does in the Z clustering, not in ZT).

The reason why the additional change inside the update() method was required is the following:  The previous version of update() did not touch swE at all (which is the essential input for Tc), only updateTc did that. Therefore, after calling update(),  swE still contained an older estimate of Tc even when updateTc was not done. Now for updating Tc, this sum must be reset to 0 before doing the summation again. However, the code I committed simply always resets swE, regardless of whether an update of Tc was requested or not (the motivation was, that  I assumed there would be no timing difference between doing this always or only doing it when requested). This is fine as long as updateTc is called before any use of Tc, but it means that the PR's update(...,false) is not completely equivalent to the previous update(). 
The difference only shows up when Tc is used after an update(false)  without making an update(true).  Before the PR, a missing updateTc simply meant that an older estimate of Tc was used instead of the most recent one. After the PR, a missing update(..,true) means that Tc is 0. So, to reproduce the old behaviour it is not enough to only change update(...,true) into update(..,false). 
This is only meant to find out why some things looked different. In a "clean" implementation it should not matter whether update(..,false) reset swE (and hence Tc) or not.
I hope this makes sense.

perrotta commented:
Thank you @werdmann for the additional explanation.

The fix is rather subtle, and it can easily get hidden within this longish but otherwise pretty technical PR. Can I suggest to implement here the "no change" version of it? I will sign it off at once as soon as the tests pass, and then you can make a short and well recognizable PR which includes only the fix, with a clearly identifiable title and description: I assume that also such additional PR can be merged quickly, as its discussion has been already brought forward here.

werdmann committed:
restore old behaviour wrt Tc updates before merging

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32764/21130

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #32764 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test
(Thank you Wolfran... even if the updates applied here does not seem to correspond to the diff you posted [yesterday](https://github.com/cms-sw/cmssw/pull/32764#issuecomment-777667645)... but let see how do the test look like)

werdmann commented:
@perotta you are right. in addition to what I posted yesterday it addresses your comment https://github.com/cms-sw/cmssw/pull/32764#discussion_r573028125 by moving the multiplication with that factor to a different loop where it needs to be executed only nvertex times instead of ntrack times.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6948be/12869/summary.html
**COMMIT**: fec4cbdbafd57b51849fb7b04f15aee70ecf2bf9
**CMSSW**: CMSSW_11_3_X_2021-02-12-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32764/12869/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-6948be/11634.911_TTbar_14TeV+2021_DD4hep+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750846
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2750813
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

perrotta commented:
In DAClusterizerInZT_vect.cc the new `update(...)` function corresponds to the previous one when `updateTc=true`. Therefore, all calls should add as argument the corresponding boolean parameter as `true`. However, I see that still L662, as well as L1211 and L1231, use the default (`false`) argument: was it done on purpose, or just forgotten?

At the contrary in DAClusterizerInZ_vect.cc all new `update()` calls now correspond to what is in the version of the code currently in CMSSW

perrotta commented:
There are a couple of differences in DQM plots for wfs 23234 and 28234. They look like just as possible numerical instabilities, but they could still be correlated with what implemented here. Eg. for wf 28234:
![image](https://user-images.githubusercontent.com/4069749/107845234-54ec4300-6dda-11eb-9072-3414fd9d635c.png)
![image](https://user-images.githubusercontent.com/4069749/107845248-67ff1300-6dda-11eb-9a23-a8adba478300.png)

werdmann commented:
> In DAClusterizerInZT_vect.cc the new `update(...)` function corresponds to the previous one when `updateTc=true`. Therefore, all calls should add as argument the corresponding boolean parameter as `true`. However, I see that still L662, as well as L1211 and L1231, use the default (`false`) argument: was it done on purpose, or just forgotten?
This is done on purpose to avoid unnecessary arithmetic when an updated Tc is not needed. The TC clusterizer only uses Tc for splitting, so it only needs to be done when a subsequent split() is possible.
> 
> At the contrary in DAClusterizerInZ_vect.cc all new `update()` calls now correspond to what is in the version of the code currently in CMSSW
Yes. In contrast to the ZT clusterizer, which did before this PR did not distinguish between update and updateTc and simply effectively did updateTc all the time, the Z clusterizer already had two versions of update and the only change was to replace update() with update(..,false) and updateTc() with update(..., true). The complication was that, there was a place where strictly speaking updateTc should have been called, but wasn't. The last commits go back to that behaviour, although it could be considered a bug (albeit one that has only small consequences if the swE sum is not reset during the update(...,false).

perrotta commented:
+1
- Technical cleanup of the c++ code
- Jenkins tests pass

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1

perrotta commented:
@werdmann : since the "technical" PR is now merged, please prepare the "fix" one based on the first IB that will contain this PR (quite likely this afternoon CMSSW_11_3_X_2021-02-15-1100)
